### PR TITLE
ScalarUDFImpl invoke improvements

### PR DESCRIPTION
**Pull request by alamb**
Builds on https://github.com/apache/datafusion/pull/13290 from @joseph-isaacs

## Which issue does this PR close?

- Follow on to  https://github.com/apache/datafusion/pull/13290
- Closes https://github.com/apache/datafusion/issues/13516

## Rationale for this change

If we are going to change the signature of `invoke` I would to make it possible in the future to reuse allocations -- at the moment the API requires creating new arrays even when the input array is not used because the functions get the input as `&[ColumnarValues]` meaning they don't own them (the caller retains a reference)

## What changes are included in this PR?

Change ScalarFunctionArgs to pass by value

1,

### CODE REVIEW

File: datafusion/physical-expr/src/scalar_function.rs [hunk 1]
```diff
@@ -140,18 +140,23 @@ impl PhysicalExpr for ScalarFunctionExpr {
...
             .collect::<Result<Vec<_>>>()?;
 
+        let inputs_all_scalars = !inputs.is_empty() && inputs
+            .iter()
+            .all(|arg| matches!(arg, ColumnarValue::Scalar(_)));
+
         // evaluate the function
-        let output = self.fun.invoke_batch(&inputs, batch.num_rows())?;
+        let output = self.fun.invoke_with_args(ScalarFunctionArgs {
+            args: inputs,
```
Comment (alamb) [inline datafusion/physical-expr/src/scalar_function.rs:149]: the whole point of this PR is to pass in `inputs` here rather than inputs.as_ref

Comment (alamb) [reply datafusion/physical-expr/src/scalar_function.rs:149]: Eventually we should probably consider doing the same thing for `evaluate` 🤔

**Comment by findepi** [comment]
> * Closes [Perf: Allow User defined functions to potentially reuse their argument arrays (to avoid new allocations) #13516](https://github.com/apache/datafusion/issues/13516)

Can we have one example function doing that, with some tests proving this indeed works?

**Comment by alamb** [comment]
> > * Closes [Perf: Allow User defined functions to potentially reuse their argument arrays (to avoid new allocations) #13516](https://github.com/apache/datafusion/issues/13516)
> 
> Can we have one example function doing that, with some tests proving this indeed works?

Just OCD following up here, there is an example / tests in this PR:
- https://github.com/apache/datafusion/pull/13637

**Pull request closed without merging**